File: src/main/kotlin/Trie.java
import java.util.HashMap;
import java.util.Map;

class Trie {

    private static class TrieNode {
        final Map<Character, TrieNode> map = new HashMap<>();
        boolean endOfWord;
    }

    private final TrieNode root;

    /**
     * Initialize your data structure here.
     */
    public Trie() {
        root = new TrieNode();
    }

    /**
     * Inserts a word into the trie.
     */
    public void insert(final String word) {
        insert(word, root);
    }

    private void insert(final String word, final TrieNode currentNode) {
        if (word == null || word.isEmpty()) return;

        final char c = word.charAt(0);

        TrieNode nextNode;

        if (currentNode.map.containsKey(c)) {
            nextNode = currentNode.map.get(c);
        } else {
            nextNode = new TrieNode();
            currentNode.map.put(c, nextNode);
        }

        if (word.length() == 1) {
            nextNode.endOfWord = true;
        } else {
            insert(word.substring(1), nextNode);
        }
    }

    /**
     * Returns if the word is in the trie.
     */
    public boolean search(final String word) {
        return search(word, root);
    }

    private boolean search(final String word, final TrieNode currentNode) {
        if (word == null || word.isEmpty()) return false;

        final char c = word.charAt(0);

        if (!currentNode.map.containsKey(c)) {
            return false;
        }

        final TrieNode nextNode = currentNode.map.get(c);

        if (word.length() == 1) {
            return nextNode.endOfWord;
        } else {
            return search(word.substring(1), nextNode);
        }
    }

    /**
     * Returns if there is any word in the trie that starts with the given prefix.
     */
    public boolean startsWith(final String prefix) {
        return startsWith(prefix, root);
    }

    private boolean startsWith(final String prefix, final TrieNode currentNode) {
        if (prefix == null || prefix.isEmpty()) return false;

        final char c = prefix.charAt(0);

        if (!currentNode.map.containsKey(c)) {
            return false;
        }

        final TrieNode nextNode = currentNode.map.get(c);

        return prefix.length() == 1 || startsWith(prefix.substring(1), nextNode);
    }
}

File: src/main/kotlin/Trie.kt
fun main(args: Array<String>) {
    val trie: Trie = TrieImpl()
    trie.add("Desert")
            .add("Dessert")
}

private interface Trie {

    fun add(word: String): Trie

    fun containsPrefix(prefix: String): Boolean

    fun containsWord(word: String): Boolean

    fun remove(word: String)
}

private class TrieNode {
    val map = mutableMapOf<Char, TrieNode>()
    var endOfWord: Boolean = false
}

private class TrieImpl : Trie {

    private val root: TrieNode = TrieNode()

    override fun add(word: String): Trie {
        if (word.isEmpty()) throw IllegalArgumentException("Word is empty!")
        add(word, root)
        return this
    }

    private fun add(word: String, currentNode: TrieNode) {
        val currentChar = word[0]

        var nextNode = TrieNode()
        if (!currentNode.map.containsKey(currentChar)) {
            nextNode.map[currentChar] = nextNode
        } else {
            nextNode = currentNode.map.getValue(currentChar)
        }

        if (word.length == 1) {
            nextNode.endOfWord = true
        } else {
            add(word.substring(1, word.length), nextNode)
        }
    }

    override fun containsPrefix(prefix: String): Boolean {
        var currentNode = root

        return containsPrefix(prefix, root)
    }

    private fun containsPrefix(prefix: String, currentNode: TrieNode): Boolean {
        val currentChar = prefix[0]

        if (currentNode.map.containsKey(currentChar)) {
            if (prefix.length == 1) {
                return true
            }

            val nextNode = currentNode.map.getValue(currentChar)
            return containsPrefix(prefix.substring(1, prefix.length), nextNode)
        }

        return false
    }

    override fun containsWord(word: String): Boolean {
        TODO("not implemented")
    }

    override fun remove(word: String) {
        TODO("not implemented")
    }
}
